dronegen/os_repos.go

```diff
@@ -0,0 +1,389 @@
+// Copyright 2021 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"path"
+	"strings"
+)
+
+func promoteBuildPipeline() []pipeline {
+	aptPipeline := promoteAptPipeline()
+	yumPipeline := promoteYumPipeline()
+	return []pipeline{
+		aptPipeline,
+		yumPipeline,
+	}
+}
+
+// Used for one-off migrations of older versions.
+// Use cases include:
+//  * We want to support another OS while providing backwards compatibility
+//  * We want to support another OS version while providing backwards compatibility
+//  * A customer wants to be able to install an older version via APT/YUM even if we
+//      no longer support it
+//  * RPM migrations after new YUM pipeline is done
+func artifactMigrationPipeline() []pipeline {
+	migrationVersions := []string{
+		// These versions were migrated as a part of the new `promoteAptPipeline`
```

Comment:
This seems a bit unclear to me given all the entries are commented out. Is it a "add the version temporarily to the list then comment it out once migrated" situation? Is there a procedure we could link to if it comes up in the future? (or does it matter?)

Reply:
That's exactly it. I can write up some docs somewhere... but outside of adding a new OS package repo type (like YUM or APK) it's pretty unlikely that we'll need to use this unless a customer complains about an old version being missing. It's really nice to have (greatly simplifies migrations) but will probably be very infrequently used after this PR is completed.